### ISSUE TITLE
Unreachable reached in collect

### ISSUE THREAD

**Issue by fredrikekre** [JuliaLang/julia#47374]
Reduced failure in a package to
```julia
function value(x)
    [value(i, x) for i in 1:1]
end
function value(i::Int, x)
    return 1.0
end
value(rand(1))
```
which gives
```
Unreachable reached at 0x7fbd8c4c4bcd

[10347] signal (4): Illegal instruction
in expression starting at REPL[3]:2
iterate at ./generator.jl:44 [inlined]
collect_to! at ./array.jl:840 [inlined]
collect_to_with_first! at ./array.jl:818 [inlined]
collect at ./array.jl:792 [inlined]
value at ./REPL[1]:2
unknown function (ip: 0x7fbd8c4c4be2)
jl_apply at /home/fredrik/julia-master/src/julia.h:1869 [inlined]
do_call at /home/fredrik/julia-master/src/interpreter.c:126
```

**Comment by fredrikekre** [comment]
git bisect points to 6baa9a60104c7e3a398b42ebb92a83007198a478 cc @Keno

**Comment by Keno** [comment]
Yeah, that commit was likely unsound. Will be undone in #47371.

**Comment by Keno** [comment]
Fixed by merge of #47371

**Comment by KristofferC** [comment]
Can we add a test for this? I'm pretty amazed that this made it through all our tests...
